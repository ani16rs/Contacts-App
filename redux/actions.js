import {login} from '../Api'

// ACTION TYPES
export const UPDATE_USER = "UPDATE_USER"
export const UPDATE_CONTACTS = "UPDATE_CONTACTS"
export const DELETE_CONTACTS = "UPDATE_CONTACTS"
export const LOG_IN_SENT = "LOG_IN_SENT"
export const LOG_IN_FULFULLED = "LOG_IN_FULFILLED"
export const LOG_IN_REJECTED = "LOG_IN_REJECTED"
export const CHANGE_FIRST_CONTACT = "CHANGE_FIRST_CONTACT"

// ACTION CREATORS
export const addUser = (newUser) => ({ type: UPDATE_USER, payload: newUser })
export const addContact = (newContact) => ({ type: UPDATE_CONTACTS, payload: newContact })
export const changeFirstContact = () => ({ type: CHANGE_FIRST_CONTACT })

// returns a function that takes a dispatch then uses that dispatch
// export const logInUser = (username, password) => async dispatch => {
//     dispatch({type: LOG_IN_SENT})
//     try {
//         const token = await login(username, password)
//         dispatch({type: LOG_IN_FULFULLED, payload: token})
//     }
//     catch (err){ dispatch({type: LOG_IN_REJECTED, payload: err.message}) }
// }

// new one with loginFn as a default arg
export const logInUser = (username, password, loginFn = login) => async dispatch => {
    // const realLoginFn = loginFn ? loginFn : logIn
    const realLoginFn = loginFn || logIn                // shorthand
    dispatch({type: LOG_IN_SENT})

    try {
        const token = await realLoginFn(username, password)
        dispatch({type: LOG_IN_FULFULLED, payload: token})
    }
    catch (err){ dispatch({type: LOG_IN_REJECTED, payload: err.message}) }
}