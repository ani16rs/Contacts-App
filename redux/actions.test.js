// const actions = require('./actions.js')
import * as actions from './actions.js'

// individual test
// test('updateUser returns the correct action', () => {
//     expect(actions.addUser({name: 'test name'})).toMatchSnapshot({ type: actions.UPDATE_USER, payload: {name: 'test name'}, })
// })

// multiple tests of same type
describe('updateUser returns action: ', () => {
    it('updateUser returns the correct action', () => {
        expect(actions.addUser({name: 'test name'})).toMatchSnapshot({ type: actions.UPDATE_USER, payload: {name: 'test name'}, })
    })

    it('handles empty object', () => {
        expect(actions.addUser({})).toMatchSnapshot()
    })

    it('handles empty name', () => {
        expect(actions.addUser({name: ''})).toMatchSnapshot()
    })
})

describe('loginUser returns actions', () => {
    const errMessage = 'shitFuckIncorrectCreds'
    const fakeToken = 'thisIsATestToken'
    const mockLogin = (username, password) => {
        if (username === 'u' && password === 'p')   return fakeToken
        throw new Error(errMessage)
    }

    it('dispatches LOG_IN_SENT', async () => {
        const mockDispatch = jest.fn()
        await actions.logInUser('', '')(mockDispatch)
        // console.log(mockDispatch.mock.calls)
        // mockDispatch.mock.calls all the args that the mock fn was invoked on
        expect(mockDispatch.mock.calls[0][0]).toEqual({type: actions.LOG_IN_SENT, })
    })

    it('dispatches LOG_IN_FULFILLED with correct creds', async () => {
        const mockDispatch = jest.fn()
        await actions.logInUser('u', 'p', mockLogin)(mockDispatch)
        expect(mockDispatch.mock.calls[1][0]).toEqual({type: actions.LOG_IN_FULFULLED, payload: fakeToken})
        expect(mockDispatch.mock.calls[1]).toMatchSnapshot()
        // console.log(mockDispatch.mock.calls)
    })

    it('dispatches LOG_IN_REJECTED with incorrect creds', async () => {
        const mockDispatch = jest.fn()
        await actions.logInUser('', '', mockLogin)(mockDispatch)
        // console.log(mockDispatch.mock.calls)
        expect(mockDispatch.mock.calls[1][0]).toEqual({type: actions.LOG_IN_REJECTED, payload: errMessage})
        expect(mockDispatch.mock.calls[1]).toMatchSnapshot()
    // expect(mockDispatch.mock.calls[1][0]).toMatchSnapshot({ payload: { name: expect.any(String), phone: expect.any(String) } })

    })
})