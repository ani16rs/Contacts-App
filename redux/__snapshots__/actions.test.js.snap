// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loginUser returns actions dispatches LOG_IN_FULFILLED with correct creds 1`] = `
Array [
  Object {
    "payload": "thisIsATestToken",
    "type": "LOG_IN_FULFILLED",
  },
]
`;

exports[`loginUser returns actions dispatches LOG_IN_REJECTED with incorrect creds 1`] = `
Array [
  Object {
    "payload": "shit fuck",
    "type": "LOG_IN_REJECTED",
  },
]
`;

exports[`updateUser returns action:  handles empty name 1`] = `
Object {
  "payload": Object {
    "name": "",
  },
  "type": "UPDATE_USER",
}
`;

exports[`updateUser returns action:  handles empty object 1`] = `
Object {
  "payload": Object {},
  "type": "UPDATE_USER",
}
`;

exports[`updateUser returns action:  updateUser returns the correct action 1`] = `
Object {
  "payload": Object {
    "name": "test name",
  },
  "type": "UPDATE_USER",
}
`;
